algosdk/abi/tuple_type.py

```diff
@@ -0,0 +1,329 @@
+        Compress consecutive boolean values into a byte for a Tuple/Array.
+        """
+        result = 0
+        if len(value_list) > 8:
+            raise error.ABIEncodingError(
+                "length of list should not be greater than 8"
+            )
+        for i, value in enumerate(value_list):
+            assert isinstance(value, bool)
+            bool_val = value
+            if bool_val:
+                result |= 1 << (7 - i)
+        return result
+
+    def encode(self, values):
+        """
+        Encodes a list of values into a TupleType ABI bytestring.
+
+        Args:
+            values (list): list of values to be encoded.
+            The length of the list cannot exceed a uint16.
+
+        Returns:
+            bytes: encoded bytes of the tuple
+        """
+        if len(self.child_types) >= (2 ** 16):
+            raise error.ABIEncodingError(
+                "length of tuple array should not exceed a uint16: {}".format(
+                    len(self.child_types)
+                )
+            )
+        tuple_elements = self.child_types
+
+        # Create a head/tail component and use it to concat bytes later
+        heads = list()
+        tails = list()
+        is_dynamic_index = list()
+        i = 0
+        while i < len(tuple_elements):
+            element = tuple_elements[i]
+            if element.is_dynamic():
+                # Head is not pre-determined for dynamic types; store a placeholder for now
+                heads.append(b"")
+                is_dynamic_index.append(True)
+                tail_encoding = element.encode(values[i])
+                tails.append(tail_encoding)
+            else:
+                if isinstance(element, BoolType):
+                    before = TupleType._find_bool(self.child_types, i, -1)
+                    after = TupleType._find_bool(self.child_types, i, 1)
+
+                    # Pack bytes to heads and tails
+                    if before % 8 != 0:
+                        raise error.ABIEncodingError(
+                            "expected before index should have number of bool mod 8 equal 0"
+                        )
+                    after = min(7, after)
+                    compressed_int = TupleType._compress_multiple_bool(
+                        values[i : i + after + 1]
+                    )
+                    # For converting one byte, the byteorder should not matter
+                    heads.append(bytes([compressed_int]))
+                    i += after
+                else:
+                    encoded_tuple_element = element.encode(values[i])
+                    heads.append(encoded_tuple_element)
+                is_dynamic_index.append(False)
+                tails.append(b"")
+            i += 1
+
+        # Adjust heads for dynamic types
+        head_length = 0
+        for head_element in heads:
+            # If the element is not a placeholder, append the length of the element
+            if head_element:
```

Comment:
This branch still does not differentiate between a 0 length head string that should be allowed (e.g. an empty tuple) and a head string that belongs to a dynamic type.

I would suggest either checking the `is_dynamic_index` list, or using `None` to represent dynamic heads and specifically checking `head_element is None` here.